Added save as functionality

jkwok678 opened the pull request.

jkwok678 committed:
Added saveToFileAsQString methods to element class

jkwok678 committed:
Renamed saveToFileAsQString as toQStringForSave in element class

jkwok678 committed:
Added toQStringForSave in the track class

jkwok678 committed:
Minimum Cmake version is now 3.12

jkwok678 committed:
Merge branch 'develop' into Add-Save-As-Functionality

jkwok678 committed:
Added a new PlatformToQString method

Used that in the code now.
Also fixed some incomplete toQStringForSave() methods.

jkwok678 committed:
Shortened ParapetTypeQString

jkwok678 committed:
Made parapetTypeQString clearer

jkwok678 committed:
Shortened StraightTypeQString

jkwok678 committed:
Shorten ParapetTypeQString

jkwok678 committed:
Shortedn StraightTrack and DirectedTrack QString

jkwok678 committed:
Shortened curved, linked and exit type strings

jkwok678 committed:
Shortened buffer, signal bridgenderpass, switch, crossover and flyover toQString strings

Also added a text file to describe all the codes.

jkwok678 committed:
Added a new test for creating the QString to write to file.

Also changed level crossing to 1 and 0 for in the toQString methods

jkwok678 committed:
Fixed ParapetToQString test

Added MapTest for directedTrackListToQString

jkwok678 committed:
Added the directedTrackListToQStringForSaving method

jkwok678 committed:
Add MapSavingCurvedTrackTest

jkwok678 committed:
Added LinkedTrackListToQString test

Also changed toQStringForSaving methods

jkwok678 committed:
Added exit and buffer track lists to QString for saving methods

Also added a test for it.

jkwok678 committed:
Added signalTrackListToQStringForSaving method

Also added aspectToQString method.
Also added a test for signalTrackListToQStringForSaving.

jkwok678 committed:
Added a levelCrossingToQString method

jkwok678 committed:
Added bridgeUnderpassTrackListToQStringForSaving

Also added a method to set Default Secondary speed and length.
Also added tests for the bridgeUnderpassTrackListToQStringForSaving method

jkwok678 committed:
Added the crossoverTrackListToQStringForSaving method

Also added test for the method.

jkwok678 committed:
Added switchTrackListTOQString for saving method

Also added a test for it.

jkwok678 committed:
Added flyoverTrackListToQStringForSaving method

Added a test for it.
Fixed default flyover secondary speed and lengths.

jkwok678 committed:
Added to the tests for StraightTrack and DirectedTrack to make it cover more scenarios

Also added a removeLevelCrossing for StraightTrack. (For debugging purposes at the moment)

jkwok678 committed:
Added more tests to cover more scenarios when saving LinkedTracks

Removed the get/setLinked methods.
Removed linked field in LinkedTrack.
Reworked classes and methods that used the old field and setter and getters.

jkwok678 committed:
Expanded tests for signalTrack/bufferTrackListToQString

jkwok678 committed:
Update mapTest.cpp

Added another scenario to BridgeUnderpassTrackToQString test

jkwok678 committed:
Added new method 'prepareFileContentToWrite()'

jkwok678 committed:
saveRailwayAs should now save StraightTrack and DirectedTrack

Also added a new filesaverTest

jkwok678 committed:
Fixed write toFileTest

jkwok678 committed:
Added CurvedTrack to the save method

jkwok678 committed:
Added LinkedTrack to prepareFileContentToWrite method

jkwok678 committed:
Fixed bug with wrong test written to file for LinkedTrack section

Added test to show that filesaver works for CurvedTrack and LinkedTrack

jkwok678 committed:
Added ExitTrack and BufferTrack to the prepareFileContentToWrite method

jkwok678 committed:
Added test for writing ExitTrack and BufferTrack

jkwok678 committed:
Added SignalTrack and BridgeUnderpassTrack to the prepareFileContentToWrite method.

jkwok678 committed:
Added a test for writing SignalTrack and BridgeUnderpassTrack.

Fixed BridgeUnderpassTrack having secondary speed and lengths.

jkwok678 committed:
Added a test for writing SwitchTrack, CrossoverTrack and FlyoverTrack

jkwok678 committed:
Added NamedLocation and Concourse to prepareFileContentToWrite method

Added a test to prepareFileContentToWrite for NamedLcoation and Concourse.
Also added concourseListToQStringForSaving and namedLocationListToQStringForSaving.

jkwok678 committed:
Added parapetListToQStringForSaving and textListToQStringForSaving methods

jkwok678 committed:
Added Parapet and Text to prepareFileContentToWrite

Added a test to check if they work.

jkwok678 committed:
Added text to front of file when saving

Added test for all components saving

jkwok678 committed:
Save As button should now work

Fixed the bridgeUnderpassTrackListToQString tests

jkwok678 committed:
Removed checking if path is empty

That's because QFileDialog doesn't let the path be empty

jkwok678 committed:
Added logic for SaveAs function

jkwok678 committed:
Added setNewFilePath method

Moved prepareFileContentToWrite to private.
Added doxygen documentation.

jkwok678 committed:
Removed unused constructor

Window now has one filesaver.
Tweaked the methods in Window to use the new filesaver methods.
Fixed filesaver tests.

jkwok678 committed:
Changed behaviour of the file menu.

Save Railway Act is only enabled after Save As is used

jkwok678 committed:
Added saving Dev2 files

jkwok678 committed:
Revert "Added saving Dev2 files"

This reverts commit 5ccc13740829fdb6952496c5c454612e21d603f6.

jkwok678 committed:
Revert "Changed behaviour of the file menu."

This reverts commit b235cb9a79849b8fb40d19d4847f2c7fbe714cf0.

jkwok678 committed:
Add save as Dev2 file

jkwok678 committed:
Save as now only works when there's at least one track on the map

zarethrex commented on canvas.cpp:
```diff
@@ -976,9 +976,14 @@ void Canvas::setGrid(bool newGrid)
+	return map->getTotalTrack();
 }
 
 void Canvas::clickCreateAddElement(ElementChosen elementToAdd, int overallX, int overallY)
```
I am not a C++ maestro but would it not be better to pass by reference here, i.e. `ElementChosen&`? Depends how you are using the object. Also if the object is not being altered consider `const ElementChosen&`.

zarethrex commented on canvas.h:
```diff
@@ -235,7 +235,13 @@ class Canvas : public QWidget
+	 * @brief Get total number of tracks.
+	 * @return tracks Number of tracks as an int.
+	 */
+	int getTrackTotal();
```
Recommend `int getTrackTotal() const;`

zarethrex replied:
(I see you did this in other places)

jkwok678 replied:
Yep probs laziness on my part, will fix.

zarethrex commented on element.cpp:
```diff
@@ -203,6 +213,17 @@ QString NamedLocation::toQString()
+QString NamedLocation::toQStringForSave()
+{
+	QString saveNamedLocationQString = locationToQString();
+	if(text != nullptr)
```
Should this not be just `if(!text)`?

jkwok678 replied:
Can't be changed, change would crash the program

zarethrex commented on element.cpp:
```diff
@@ -229,6 +250,17 @@ QString Concourse::toQString()
+QString Concourse::toQStringForSave()
+{
+	QString saveConcourseQString = locationToQString();
+	if(text != nullptr)
```
Same as above

zarethrex commented on filesaver.cpp:
```diff
@@ -1,34 +1,85 @@
-Filesaver::Filesaver(QString newPath)
+//private
+
+std::vector<QString> Filesaver::prepareFileContentToWrite(Map *map)
```
As `map` is not altered, maybe `const Map* map`?

zarethrex commented on filesaver.h:
```diff
@@ -9,11 +9,36 @@ class Filesaver
+	 * @brief Set a new file path for the filesaver.
+	 * @param newPath QString path.
+	 */
+	void setNewFilePath(QString newPath);
```
Again I cannot claim to be an expert but I would use `const QString&`

zarethrex commented on filesaver.h:
```diff
@@ -9,11 +9,36 @@ class Filesaver
+	 * @param map The current map on screen.
+	 * @return true if it saved, false if it failed to save.
+	 */
+	bool saveRailway(Map *map);
```
Is `map` altered here? Else `const Map*`

zarethrex commented on track.h:
```diff
@@ -704,16 +754,10 @@ class LinkedTrack : public Track
+	 * @return true if it's linked otherwise false.
 	 */
-	void setLinked(bool newLinked);
+	bool isLinked();
```
`bool isLinked() const;` matching others

zarethrex commented on window.cpp:
```diff
@@ -143,14 +144,31 @@ void Window::loadRailway()
+		if (drawingArea->checkAllTracksInMapConnected())
+		{
+			fileName = QFileDialog::getSaveFileName(this, tr("Save File"),
+								   "./",
```
Suggest not using hard coded paths, for current directory: `QDir::currentPath()`

zarethrex commented on window.cpp:
```diff
@@ -143,14 +144,31 @@ void Window::loadRailway()
+		else
+		{
+			fileName = QFileDialog::getSaveFileName(this, tr("Save File"),
+								   "./",
```
See above

jkwok678 committed:
Quality of code changes

In canvas, use const in getTotalTrack.
In element toQString methods use (if!text).
Add const to saveRailway and prepareFileContentToWrite in filesaver
In LinkedTrack islinked is now const.
Use currentPath instead of "./" in when opening filedialog

jkwok678 committed:
Revert "Quality of code changes"

This reverts commit 1ce6ba20ac70cd3340644491495404be92dd213c.

jkwok678 committed:
Code quality improved

Const added to canvas getTrackTotal.
Const added to linkedTrack isLinked.
QFileDialog in window has "./" replaced with QDir::currentPath()

jkwok678 committed:
Add Track NamedLocation

jkwok678 committed:
Added methods to check if an element or track is at a location

jkwok678 committed:
Added more tests to MapSavingQStringTest

Added namedLocationListToQString.
Added concourseToQString.

jkwok678 committed:
Added tests for all elements getting them from the map with coordinates.

jkwok678 committed:
Added tests to check if the check element exists methods work

jkwok678 committed:
Changed Implementation of checkElementExists method

jkwok678 committed:
Merge branch 'develop' into Add-Save-As-Functionality

jkwok678 committed:
Added new method in preperation for adding named location to track.

jkwok678 committed:
Add new linkNewElementToText methods

Unfinished right now.

jkwok678 committed:
Removed QMessageBox when user clicks on empty space.

jkwok678 committed:
Remove showElementAlreadyThereErrorMessage method from Map class

jkwok678 committed:
Revert "Remove showElementAlreadyThereErrorMessage method from Map class"

This reverts commit e7c3d57b17439f36f4cd2965922019ce1c6157fd.

jkwok678 committed:
Revert "Removed QMessageBox when user clicks on empty space."

This reverts commit 84d6e6ea0b03f9dcc7e73b61024171546e0956e3.

jkwok678 committed:
Removed error message when user clicks on empty space.

jkwok678 committed:
Removed showElementAlreadyThereErrorMessage error message.

jkwok678 committed:
Convert showAllTrackConnecedSuccessMessage to use object instead.

jkwok678 committed:
showElementAlreadyThereErrorMessage is no longer static

jkwok678 committed:
Window class changed to use new non-static method.

jkwok678 committed:
showLowNumOfLinkedTrackErrorMessage is no longer static

jkwok678 committed:
showOddNumOfLinkedTrackErrorMessage is no longer static

jkwok678 committed:
showNotAllTrackConnectedErrorMessage is no longer static

jkwok678 committed:
showZeroTrackErrorMessage

jkwok678 committed:
ShowZeroTrackErrorMessage is not non-static.

jkwok678 committed:
Merge branch 'Add-Non-Static-Messages' of https://github.com/jkwok678/RailwayBuilder into Add-Non-Static-Messages

jkwok678 committed:
Merge pull request #7 from jkwok678/Add-Non-Static-Messages

Message class no longer relies on static methods.

jkwok678 committed:
Add ability for new elements to connect to text that is local.

jkwok678 committed:
Deleting an element with a name also deletes the text.

jkwok678 committed:
Fix totalTrack not updating when ExitTrack is added.

jkwok678 committed:
Add ability to add namedLocation to certain tracks.

Issues:
GetElement returns only one element so far.
Track and NamedLocation is seperate right now, Track doesn't know it is named or has the potential to be named.

jkwok678 committed:
Stop ability to add NamedLocation to tracks with Platforms or Level Crossings

jkwok678 committed:
Rename getter and setter for Track's LinkedNamedLocation

jkwok678 committed:
Fix addLevelCrossing decision logic

jkwok678 committed:
Add Track link to NamedLocation when adding NamedLocation to track

jkwok678 committed:
Fix bug that allows platforms to be placed on NamedLocations

jkwok678 committed:
Fix a bug where a level crossing can be added onto a NamedLocation

jkwok678 committed:
Make NamedElements easier to read

jkwok678 committed:
Change toQString and QStringForSave methods for NamedLocation

They now just save the link to the text instead of the whole text object,

jkwok678 committed:
Add test to check if NamedLocation is saved correctly with text

jkwok678 committed:
toQString shows whole text object and toQStringForSaving doesn't

Tests have been tweaked for NamedLocations and Concourses to match new code.

jkwok678 committed:
Change toQString and toQStringForSave methods

All toQString mehtods use text->toQString and toQStringForSave uses text->getReadableText

jkwok678 committed:
Add to mapTest

jkwok678 committed:
Fux failing curvedTrackListToQString

jkwok678 committed:
Update build-test.yml

jkwok678 committed:
Merge branch 'develop' into Add-Save-As-Functionality
